src/pint/toa.py

```diff
@@ -1917,3 +1920,73 @@ def read_toa_file(self, filename, process_includes=True, top=True):
                         newtoa.flags["to"] = self.cdict["TIME"]
                     self.toas.append(newtoa)
                     ntoas += 1
+
+
+def merge_TOAs(TOAs_list):
+    """Merge a list of TOAs instances and return a new combined TOAs instance
+
+    In order for a merge to work, each TOAs instance needs to have
+    been created using the same Solar System Ephemeris (EPHEM),  
+    the same reference timescale (i.e. CLOCK), and the same value of
+    .planets (i.e. whether planetary PosVel columns are in the tables
+    or not).
+
+    Parameters
+    ----------
+    TOAs_list : list of TOAs instances
+
+    Returns
+    -------
+    A new TOAs instance with all the combined and grouped TOAs
+    """
+    # Check each TOA object for consistency
+    ephems = [tt.ephem for tt in TOAs_list]
+    if len(set(ephems)) > 1:
+        log.warning(f"EPHEMs are inconsistent in merge_TOAs(): {ephems}\nNot merging.")
+        return None
+    inc_BIPM = [tt.clock_corr_info["include_bipm"] for tt in TOAs_list]
+    if len(set(inc_BIPM)) > 1:
+        log.warning(
+            f"include_bipms are inconsistent in merge_TOAs(): {inc_BIPM}\nNot merging."
+        )
+        return None
+    BIPM_vers = [tt.clock_corr_info["bipm_version"] for tt in TOAs_list]
+    if len(set(BIPM_vers)) > 1:
+        log.warning(
+            f"bipm_versions are inconsistent in merge_TOAs(): {BIPM_vers}\nNot merging."
+        )
+        return None
+    inc_GPS = [tt.clock_corr_info["include_gps"] for tt in TOAs_list]
+    if len(set(inc_GPS)) > 1:
+        log.warning(
+            f"include_gpss are inconsistent in merge_TOAs(): {inc_GPS}\nNot merging."
+        )
+        return None
+    planets = [tt.planets for tt in TOAs_list]
+    if len(set(planets)) > 1:
+        log.warning(
+            f"planets are inconsistent in merge_TOAs(): {planets}\nNot merging."
+        )
+        return None
+    num_cols = [len(tt.table.columns) for tt in TOAs_list]
+    if len(set(num_cols)) > 1:
+        log.warning(
+            f"Numbers of table columns are inconsistent in merge_TOAs(): {num_cols}\nNot merging."
+        )
+        return None
+    # Use a copy of the first TOAs instance as the base for the joined object
+    nt = copy.deepcopy(TOAs_list[0])
+    nt.filename = [tt.filename for tt in TOAs_list]
```

Comment:
Can this merge TOAs objects that have lists of filenames?

Reply:
Good point.  I think that will be rare, but I'll implement that today.